flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java

```diff
@@ -204,6 +212,10 @@ protected void updateTaskExecutionStateInternal(
             final ExecutionVertexID executionVertexId,
             final TaskExecutionStateTransition taskExecutionState) {
 
+        if (taskExecutionState.getExecutionState() == ExecutionState.FINISHED) {
+            stopReserveAllocation(executionVertexId);
+        }
```

Comment:
It is a bit unfortunate that the contract between `stopReserveAllocation` and releasing of a slot is a bit implicit via the `ExecutionState.FINISHED`. It would be more explicit if we had a `releaseAssignedResource` call in the `DefaultScheduler` but the release logic is hidden in the depth of the `ExecutionGraph`. Maybe we can add a comment explaining this situation.

Reply:
Yes, I added some comments to explain it.

Reply:
Done.